Spigot-Server-Patches/0264-Optimize-Hoppers-with-Plugin-Heuristics.patch

```diff
@@ -0,0 +1,428 @@
++        final int origCount = itemstack.getCount();
++        final World world = ihopper.getWorld();
++        final int moved = Math.min(world.spigotConfig.hopperAmount, origCount);
++        itemstack.setCount(moved);
++        final ItemStack itemstack2 = addItem(iinventory, ihopper, itemstack, null);
++        if (itemstack2.isEmpty()) {
++            itemstack = itemstack.cloneItemStack();
++            itemstack.setCount(origCount - moved);
++            IGNORE_TILE_UPDATES = true;
++            iinventory.setItem(i, itemstack);
++            IGNORE_TILE_UPDATES = false;
++            iinventory.update();
++            return true;
++        }
++        itemstack.setCount(origCount);
++        return true;
++    }
++
++    // Mode 2 = Only 1 event fire, no cloning, no plugins use getItem/setItem
++    private static boolean pullMode2(IHopper ihopper, IInventory iinventory, int i) {
++        ItemStack itemstack = iinventory.getItem(i);
++        final int origCount = itemstack.getCount();
++        final World world = ihopper.getWorld();
++        final int moved = Math.min(world.spigotConfig.hopperAmount, origCount);
++        itemstack.setCount(moved);
++
++        if (!pullMode2EventFired) {
++            pullMode2EventFired = true;
++            Inventory sourceInventory;
++            // Have to special case large chests as they work oddly
++            if (iinventory instanceof InventoryLargeChest) {
++                sourceInventory = new org.bukkit.craftbukkit.inventory.CraftInventoryDoubleChest((InventoryLargeChest) iinventory);
++                // Paper start - avoid redundant snapshot creation of a TE
++            } else if (iinventory instanceof TileEntity) {
++                sourceInventory = ((TileEntity) iinventory).getOwner(false).getInventory();
++            } else {
++                sourceInventory = iinventory.getOwner().getInventory();
++            }
++
++            Inventory destination;
++            if (ihopper instanceof TileEntity) {
++                destination = ((TileEntity) ihopper).getOwner(false).getInventory();
++            } else {
++                destination = ihopper.getOwner().getInventory();
++            }
++
++            InventoryMoveItemEvent event = new InventoryMoveItemEvent(sourceInventory,
++                    // Mirror is safe as we no plugins ever use this item
++                    CraftItemStack.asCraftMirror(itemstack), destination, false);
++            if (!event.callEvent()) {
++                itemstack.setCount(origCount);
++                if (ihopper instanceof TileEntityHopper) {
++                    ((TileEntityHopper) ihopper).setCooldown(ihopper.getWorld().spigotConfig.hopperTransfer);
++                } else if (ihopper instanceof EntityMinecartHopper) {
++                    ((EntityMinecartHopper) ihopper).setCooldown(ihopper.getWorld().spigotConfig.hopperTransfer / 2);
++                }
++                return false;
```

Comment:
note to self, change this to true. false causes this code to be ran for every item. but because we are mode 2, we know item is not going to change the fact this event was cancelled.

return true in mode 0 too. i was originally concerned it would block plugins who want to cancel first and allow 2nd, but the same cant be said for push mode, so we can just make them consistent, and greatly improve performance of cancelled pulls.